ps3.py

```diff
@@ -0,0 +1,401 @@
+      from the hand.
+
+    * An invalid word is rejected, and a message is displayed asking
+      the user to choose another word.
+
+    * After every valid word: the score for that word is displayed,
+      the remaining letters in the hand are displayed, and the user
+      is asked to input another word.
+
+    * The sum of the word scores is displayed when the hand finishes.
+
+    * The hand finishes when there are no more unused letters.
+      The user can also finish playing the hand by inputing two 
+      exclamation points (the string '!!') instead of a word.
+
+      hand: dictionary (string -> int)
+      word_list: list of lowercase strings
+      returns: the total score for the hand
+      
+    """
+    score = 0
+    while hand != {}:
+        print('\nCurrent hand: ', end='')
+        display_hand(hand)
+        word = input('Enter word, or “!!” to indicate that you are finished: ')
+        if word == '!!':
+            break
+        elif is_valid_word(word, hand, word_list):
+            score += get_word_score(word, HAND_SIZE)
+            print('"' + word + '" earned', get_word_score(word, HAND_SIZE), 'points. Total:', score, 'points')
+        else:
+            print('This is not a valid word.', end='')
+            if update_hand(hand, word) != {}:
+                print(' Please choose another word.')
+            else:
+                print()
+        hand = update_hand(hand, word)
+    if hand == {}:
+        print('Ran out of letters')
+    print('Total score for this hand:', score, 'points')
+    return score
+
+
+#
+# Problem #6: Playing a game
+# 
+
+
+#
+# procedure you will use to substitute a letter in a hand
+#
+
+def substitute_hand(hand, letter):
+    """ 
+    Allow the user to replace all copies of one letter in the hand (chosen by user)
+    with a new letter chosen from the VOWELS and CONSONANTS at random. The new letter
+    should be different from user's choice, and should not be any of the letters
+    already in the hand.
+
+    If user provide a letter not in the hand, the hand should be the same.
+
+    Has no side effects: does not mutate hand.
+
+    For example:
+        substitute_hand({'h':1, 'e':1, 'l':2, 'o':1}, 'l')
+    might return:
+        {'h':1, 'e':1, 'o':1, 'x':2} -> if the new letter is 'x'
+    The new letter should not be 'h', 'e', 'l', or 'o' since those letters were
+    already in the hand.
+    
+    hand: dictionary (string -> int)
+    letter: string
+    returns: dictionary (string -> int)
+    """
+    type_letter = ""
+    if letter in hand.keys():
+        new_hand = hand.copy()
+        if letter in VOWELS:
+            type_letter = VOWELS
+        elif letter in CONSONANTS:
+            type_letter = CONSONANTS
+        while True:
+            new_letter = random.choice(type_letter)
+            if new_letter not in hand.keys():
+                new_hand[new_letter] = new_hand.pop(letter)
+                return new_hand
+    return hand
```

Comment:
This code doesn't work right. Change this function according the rules of the game, which were presented in docstrings. You have a possibility to choose a letter from all letters which aren't presented in hand.